split initialization of sonic_platform to a per-daemon basis

stephenxs opened the pull request:
**- What I did**
split sonic_platform initialization to a per-daemon basis to save the startup time of daemons
make Mellanox new platform api able to be installed on host so that reboot cause can be handled on it.

**- How I did it**
1. modify makefiles so that the following modules are installed on host:
    - sonic_platform, which handles reboot cause
    - sonic_daemon_base, on which sonic_platform depends
2. remove unnecessary dependencies of sonic_platform on host
    - implement platform.py
    - add an added argument to sonic_platform.platform.Platform's constructor, as following:
    `Platform.__init__(self, daemon = None)`
      - [TBD] the base class sonic_platform_base.platform_base.PlatformBase.__init__ should be updated accordingly
    - Platform's constructor loads and initializes only the functionalities on a per-daemon basis, i.e. only functionalities required by the daemon are initialized for it.
    - define initializers for each components belonging the chassis and remove them out of the chassis's constructor so that they can be called only when needed. currently the following initializers defined:
      - initialize_psu
      - initialize_watchdog
      - initialize_fan
      - initialize_sfp
      - initialize_eeprom
      - initialize_components_list

**- How to verify it**
load chassis on host and check whether reboot-cause and watchdog can be well handled.
load chassis on pmon docker and check whether other components can be well handled.
run process-reboot-cause and check whether the correct reboot cause can be parsed.


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -982,13 +1119,32 @@ def get_tx_power(self):
+        # Reset SFP
+        pmaos.rst = 1
+        meta.access_cmd = SXD_ACCESS_CMD_SET
+        rc = sxd_access_reg_pmaos(pmaos, meta, 1, None, None)
```
change to REGISTER_NUM

stephenxs replied:
fixed

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -982,13 +1119,32 @@ def get_tx_power(self):
+        meta = self._init_sx_meta_data()
+        meta.access_cmd = SXD_ACCESS_CMD_GET
+
+        rc = sxd_access_reg_pmaos(pmaos, meta, 1, None, None)
```
change to REGISTER_NUM

stephenxs replied:
fixed

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -732,7 +842,23 @@ def get_lpmode(self):
+            meta = self._init_sx_meta_data()
+            meta.access_cmd = SXD_ACCESS_CMD_GET
+
+            rc = sxd_access_reg_mcion(mcion, meta, 1, None, None)
```
change to REGISTER_NUM

stephenxs replied:
fixed

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -982,13 +1119,32 @@ def get_tx_power(self):
-    
+        handle = self._open_sdk()
+        if handle is None:
+            return None
```
return false?

stephenxs replied:
fixed

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -741,8 +867,19 @@ def get_power_override(self):
+            offset = 0
+            sfpd_obj = sff8436Dom()
+            if sfpd_obj is None:
+                return None
```
return value?

stephenxs replied:
fixed

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -732,7 +842,23 @@ def get_lpmode(self):
+            lpm_mask = 1 << 8
+            lpm_status = (lpm_mask & mcion.module_status_bits) != 0
+
+            return lpm_status
```
boolean?

stephenxs replied:
sure. it is a boolean.

keboliu commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -723,7 +826,14 @@ def get_tx_disable_channel(self):
+        for i in range(len(tx_disable_list)):
+            if tx_disable_list[i]:
+                tx_disabled |= 1 << i
+        return tx_disabled
```
boolean?

stephenxs replied:
no. it's sth like a bit string.

Commit:
install new platform api on host

Commit:
1. install necessary dependencys of chassis api on host, like sonic-daemon-base
2. remove dependency on the modules not feasible to install on host, like swsscomm, sdk, etc.

Commit:
add dependency on SONIC_PLATFORM_API_PY2 for SONIC_INSTALLERS

Commit:
fix type in process-reboot-cause

Commit:
fix the issue "'NoneType' object has no attribute 'closelog'" which results from referencing an already destroyed global variable when destructuring daemon_base.Logger by introducing a class member to represent the global

Commit:
enhance the reboot cause by supporting aux power failure, long/short push on power button

Commit:
optimize chassis's initialization by
1. removing iniaializations of varient components out from constructor and providing one initializer for each.
2. only import other modules when initializing it.
by doing so, initializing time is reduced and unnecessary dependencies of chassis is removed for daemons who don't need them.

Commit:
[sonic_platform]
we decide to split chassis in a host/daemon basis rather than a per-daemon bases.
adjust the argument of Platform() accordingly as well as all code that calls Platform.

stephenxs commented:
I'm going to split this PR into 3 which contain the following logic and then close it.
1. installing vendor-specific sonic_platform on the host:
    - make file rules which contains the logic of making it and installing it.
    - remove unnecessary dependencies which can block sonic_platform from installing on the host.
2. fix the sonic_daemon_base's issue which reports error when destructing sonic_daemon_base.
3. mellanox specific part:
    - split initializers of variant categories of devices so that unnecessary dependencies can be avoided
    - improve reboot-cause-handling logic
